agent-connection/src/main/kotlin/com/luxoft/blockchainlab/corda/hyperledger/indy/PythonRefAgentConnection.kt

```diff
@@ -57,137 +94,174 @@ class PythonRefAgentConnection : AgentConnection {
-        log.info { "AgentConnection to $serverUri closed: $code,$reason,$remote" }
-    }
-
-    val receivedMessages = mutableListOf<String>()
-
-    override fun onMessage(message: String?) {
-        log.info { "Message from $serverUri: $message" }
-        if (message != null)
-            synchronized(receivedMessages) { receivedMessages.add(message) }
-    }
-
-    override fun onError(ex: Exception?) {
-        log.warn(ex) { "AgentConnection to $serverUri error" }
-    }
-
-    fun sendJson(obj: Any) = send(SerializationUtils.anyToJSON(obj))
-
-    fun popMessageOfType(type: String): String? {
-        synchronized(receivedMessages) {
-            val result = receivedMessages.find { SerializationUtils.jSONToAny<Map<String, Any>>(it)["@type"].toString().contentEquals(type) }
-            if (result != null)
-                receivedMessages.remove(result)
-
-            return result
+    private fun sendAsJson(obj: Any) = webSocket.sendAsJson(obj)
+
+    /**
+     * Retrieves a new invite from the agent
+     */
+    override fun generateInvite(): Single<String> {
+        return Single.create { observer ->
+            try {
+                /**
+                 * to generate the invite, send "generate_invite" message to the agent, and wait for "invite_generated"
+                 * which must contain an invite
+                 */
+                webSocket.sendAsJson(SendMessage(type = MESSAGE_TYPES.GENERATE_INVITE))
+                webSocket.receiveMessageOfType<ReceiveInviteMessage>(MESSAGE_TYPES.INVITE_GENERATED).subscribe({ msg ->
+                    observer.onSuccess(msg.invite)
+                }, { e: Throwable -> throw(e) })
+            } catch (e: Throwable) {
+                observer.onError(e)
+            }
         }
     }
 
-    fun waitForMessageOfType(type: String): String {
-        var messageOfType: String? = null
-        while (messageOfType == null) {
-            sleep(500)
-            messageOfType = popMessageOfType(type)
+    /**
+     * Waits for incoming connection from whatever party which possesses the invite
+     */
+    override fun waitForInvitedParty(invite: String): Single<IndyPartyConnection> {
+        return Single.create { observer ->
+            try {
+                if (getConnectionStatus() == AgentConnectionStatus.AGENT_CONNECTED) {
+                    /**
+                     * On an incoming connection request, the agent must send the "request_received"
+                     */
+                    webSocket.receiveMessageOfType<RequestReceivedMessage>(MESSAGE_TYPES.REQUEST_RECEIVED).subscribe({
+                        /**
+                         * on the "request_received" message, reply with "send_response" with remote DID to any incoming connection request.
+                         * at this stage it doesn't matter which party sent the connection request
+                         */
+                        sendAsJson(RequestSendResponseMessage(it.did))
```

Comment:
Invite processing will be dropped if end-client came with delay.

Reply:
The logic need to be split on 2 functions. The pooling process has to be moved to separated routing loop and wakes up when new consumer comes, otherwise it sleeps almost time.

```
waitForInvitedParty(invite) {
    counter++
    waitMessageOfType(MESSAGE_TYPE.PAIRWISE_ESTABLISHED, pubKey).subscribe {
        //TODO new indy connection
        counter--
    }

}
```

```
poolingLoop {
  while(counter > 0) {
      sendAsJson(StateRequest())
      webSocket.receiveMessageOfType<ObjectNode>(MESSAGE_TYPES.STATE_RESPONSE).subscribe( {
               parsePairwise
                    notifyByPublicKey
      })
      receiveMessageOfType<...>(MESSAGE_TYPES.REQUEST_RECIEVED).subscribe( {
          sendAsJson(did)
          receiveMessageOfType<...>(MESSAGE_TYPES.REQUEST_STATE_RESPONSE).subscribe( {
          // TODO: just log
      }
   })
  }
}
```

Reply:
done.

Reply:
see 32bfa1288a311597f1be1131ec9bdf2db7b282bd

Reply:
finally b0137a1355fa8e6e2ed9664591562ce72452e8bd

Reply:
4cea462199bf1868df9ead760d9ddf35a8653a45

Reply:
ed75ba9e5ed3216cee38126f4fe84dca1137bb19

Reply:
Could we make in that way

fetchAgentData has to be run in separated thread and be suspended when no active connections. Otherwise we block all threads pollStatusTo.. , Im afraid it may prevent to finally blocked execution pool.

```
val numberOfPendingInvites = AtomicInteger()

    private fun fetchAgentData() {
        while (numberOfPendingInvites.get() > 0) {
            sendAsJson(StateRequest())
            webSocket.receiveMessageOfType<ObjectNode>(MESSAGE_TYPES.STATE_RESPONSE).subscribe { agentData ->
                agentData["content"]["pairwise_connections"].forEach { pairwise ->
                    val publicKey = pairwise["metadata"]["connection_key"].asText()

                    if (awaitingPairwiseConnections.contains(publicKey))
                        awaitingPairwiseConnections.remove(publicKey)!!.onSuccess(pairwise)
                }
            }
        }
    }


    fun waitForInvitedParty(invite: String): Single<IndyPartyConnection> {
        return Single.create { observer ->
            try {
                if (getConnectionStatus() != AgentConnectionStatus.AGENT_CONNECTED)
                    throw AgentConnectionException("Agent is disconnected")

                val waitForPairwise = Single.create<JsonNode>({ client ->
                    numberOfPendingInvites.incrementAndGet()
                    awaitingPairwiseConnections[getPubkeyFromInvite(invite)] = client
                })

                waitForPairwise
                        .timeout(60000, TimeUnit.MILLISECONDS)
                        .subscribe({ pairwise ->
                            val indyParty = IndyParty(webSocket,
                                    pairwise["their_did"].asText(),
                                    pairwise["metadata"]["their_endpoint"].asText(),
                                    pairwise["metadata"]["their_vk"].asText(),
                                    pairwise["my_did"].asText())
                            indyParties[pairwise["their_did"].asText()] = indyParty
                            observer.onSuccess(indyParty)
                        }, { e: Throwable -> throw(e) } )

            } catch (e: Throwable) {
                observer.onError(e)
            } finally {
                numberOfPendingInvites.decrementAndGet()
            }
        }
    }
```

Reply:
done